lib/update/cluster/phases/system.go

```diff
@@ -122,50 +126,39 @@ type updatePhaseConfig struct {
 	ServiceUser storage.OSUser
 	// ExecutorParams is common phase executor parameters
 	fsm.ExecutorParams
-	// FieldLogger allows logging
-	logrus.FieldLogger
+	// FieldLogger specifies the logger
+	log.FieldLogger
 	// remote is the remote executor
 	remote fsm.Remote
 }
 
 // NewUpdatePhaseConfig returns executor that pulls local configuration packages
-func NewUpdatePhaseConfig(c FSMConfig, p fsm.ExecutorParams, remote fsm.Remote) (*updatePhaseConfig, error) {
-	cluster, err := c.Operator.GetLocalSite()
+func NewUpdatePhaseConfig(
+	p fsm.ExecutorParams,
+	operator ops.Operator,
+	packages pack.PackageService,
+	localPackages pack.PackageService,
+	remote fsm.Remote,
+	logger log.FieldLogger,
+) (*updatePhaseConfig, error) {
+	cluster, err := operator.GetLocalSite()
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 	return &updatePhaseConfig{
-		Packages:       c.ClusterPackages,
-		LocalPackages:  c.HostLocalPackages,
+		Packages:       packages,
+		LocalPackages:  localPackages,
 		ServiceUser:    cluster.ServiceUser,
 		ExecutorParams: p,
-		FieldLogger:    logrus.NewEntry(logrus.New()),
+		FieldLogger:    logger,
 		remote:         remote,
 	}, nil
 }
 
 // Execute pulls rotated teleport master config package to the local package store
-func (p *updatePhaseConfig) Execute(context.Context) error {
-	update, err := pack.FindLatestPackageWithLabels(
-		p.Packages, p.Plan.ClusterName, map[string]string{
-			pack.AdvertiseIPLabel: p.Phase.Data.Server.AdvertiseIP,
-			pack.OperationIDLabel: p.Plan.OperationID,
-			pack.PurposeLabel:     pack.PurposeTeleportMasterConfig,
-		})
-	if err != nil && !trace.IsNotFound(err) {
-		return trace.Wrap(err)
-	}
-	if trace.IsNotFound(err) {
-		p.Info("No teleport master config update found.")
-		return nil
-	}
-	p.Infof("Pulling teleport master config update: %v.", update)
-	_, err = service.PullPackage(service.PackagePullRequest{
-		SrcPack: p.Packages,
-		DstPack: p.LocalPackages,
-		Package: *update,
-		Upsert:  true,
-	})
+func (p *updatePhaseConfig) Execute(ctx context.Context) error {
+	b := utils.NewExponentialBackOff(5 * time.Minute)
+	err := utils.RetryTransient(ctx, b, p.pullUpdates)
```

Comment:
Need to retry in case of connection failure to account for the use-case when this phase runs after the etcd upgrade which restarts the `gravity-site`.